crates/transaction-generator-lib/src/p2p_transaction_generator.rs

```diff
@@ -7,35 +7,108 @@ use aptos_sdk::{
-use rand::{
-    distributions::{Distribution, Standard},
-    prelude::SliceRandom,
-    rngs::StdRng,
-    Rng, RngCore, SeedableRng,
-};
+use rand::{distributions::{Distribution, Standard}, prelude::SliceRandom, rngs::StdRng, Rng, RngCore, SeedableRng};
 use std::{cmp::max, sync::Arc};
+use std::cmp::min;
+
+
+/// Specifies how to get a given number of samples from an item pool.
+trait Sampler: Send + Sync {
+    fn sample(&mut self, rng: &mut StdRng, count: usize) -> Vec<usize>;
+}
+
+/// A sampler that samples a random subset of the pool. Samples are replaced immediately.
+struct BasicSampler {
+    pool_size: usize,
+}
+
+impl Sampler for BasicSampler {
+    fn sample(&mut self, rng: &mut StdRng, count: usize) -> Vec<usize> {
+        rand::seq::index::sample(rng, self.pool_size, count).into_vec()
+    }
+}
+
+
+/// A samplers designed for generating a block of P2P transfers without read-write conflicts.
+/// The pool is divided into sub-pools, one of the them being the primary:
+/// it will keep serving the sample requests *without replacement*, until it's depleted.
+/// When the current primary is depleted, another sub-pool takes over and the current one resets.
+struct BurnAndRecycleSampler {
+    /// We store all sub-pools together in 1 Vec: `item_pool[segment_size * x..segment_size * (x+1)]` being the x-th sub-pool.
+    item_pool: Vec<usize>,
+    next_index: usize,
+    segment_size: usize,
+    init_shuffle_done: bool
+}
+
+impl BurnAndRecycleSampler {
+    fn new(num_items: usize, segment_size: usize) -> Self {
+        Self {
+            item_pool: (0..num_items).collect(),
+            next_index: 0,
+            segment_size,
+            init_shuffle_done: false,
+        }
+    }
+
+    fn sample_one(&mut self, rng: &mut StdRng) -> usize {
+        if !self.init_shuffle_done {
+            self.item_pool.shuffle(rng);
+            self.init_shuffle_done = true;
+        }
+        if self.next_index % self.segment_size == 0 {
+            // Switching to a new sub-pool: shuffle it first.
+            let segment_end = min(self.item_pool.len(), self.next_index + self.segment_size);
+            self.item_pool[self.next_index..segment_end].shuffle(rng);
+        }
+        let sampled = self.item_pool[self.next_index];
+        self.next_index = (self.next_index + 1) % self.item_pool.len();
+        sampled
+    }
+}
+
+impl Sampler for BurnAndRecycleSampler {
+    fn sample(&mut self, rng: &mut StdRng, count: usize) -> Vec<usize> {
+        (0..count).map(|_| self.sample_one(rng)).collect()
+    }
+}
+
+#[test]
+fn test_burn_and_recycle_sampler() {
+    let mut rng = StdRng::from_entropy();
+    let mut sampler = BurnAndRecycleSampler::new(6, 3);
+    for _ in 0..20 {
+        let s = sampler.sample_one(&mut rng);
+        println!("s={s}");
+    }
+}
+
 
 pub struct P2PTransactionGenerator {
     rng: StdRng,
     send_amount: u64,
     txn_factory: TransactionFactory,
     all_addresses: Arc<RwLock<Vec<AccountAddress>>>,
```

Comment:
also sampling by index is generally fine, but code is made so that these Arc's are communication mechanisms between transaction generators.

i.e:
- we can have first phase of account creation that populates all_addresses, and second phase of coin transfers to consume it
- or you can have a mix of account creation that populate all_addresses, and p2p transfers that consumes it. 
- or you can first populate all_addresses from account creation, and then run a mix of p2p transfers and another transaction type that consumes it

so all_addresses can change

it might be a bit nicer with the rest of the system, to remove accounts from all_addresses, and then place them back later (instead of just operating based on index, initially set)

Reply:
@igor-aptos which one sounds better, to change P2PTxnGenerator to hold the burnt address, or to augment the shared pool to have a burnt queue or something?